From b1926af75885d661f07a7b4c6baaa587cbbc78b0 Mon Sep 17 00:00:00 2001
From: Eric Sandeen <sandeen@redhat.com>
Date: Mon, 4 Aug 2014 19:03:48 -0400
Subject: [fs] btrfs: mark mapping with error flag to report errors to userspace

Message-id: <1407179109-13714-342-git-send-email-sandeen@redhat.com>
Patchwork-id: 87176
O-Subject: [RHEL7.1 PATCH 341/422] Btrfs: mark mapping with error flag to report errors to userspace
Bugzilla: 1071179
RH-Acked-by: Zach Brown <zab@redhat.com>

From: Liu Bo <bo.li.liu@oracle.com>

commit 5dca6eea91653e9949ce6eb9e9acab6277e2f2c4
Author: Liu Bo <bo.li.liu@oracle.com>
Date:   Mon May 12 12:47:36 2014 +0800

    Btrfs: mark mapping with error flag to report errors to userspace

    According to commit 865ffef3797da2cac85b3354b5b6050dc9660978
    (fs: fix fsync() error reporting),
    it's not stable to just check error pages because pages can be
    truncated or invalidated, we should also mark mapping with error
    flag so that a later fsync can catch the error.

    Signed-off-by: Liu Bo <bo.li.liu@oracle.com>
    Signed-off-by: Chris Mason <clm@fb.com>
Signed-off-by: Eric Sandeen <sandeen@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/btrfs/extent_io.c b/fs/btrfs/extent_io.c
index 498b68b..af0cd6b 100644
--- a/fs/btrfs/extent_io.c
+++ b/fs/btrfs/extent_io.c
@@ -2367,6 +2367,8 @@ int end_extent_writepage(struct page *page, int err, u64 start, u64 end)
  if (!uptodate) {
   ClearPageUptodate(page);
   SetPageError(page);
+  ret = ret < 0 ? ret : -EIO;
+  mapping_set_error(page->mapping, ret);
  }
  return 0;
 }
-- 
1.7.1